ethereum_history_api/circuits/lib/src/verifiers/receipt.nr

```diff
@@ -1,43 +1,67 @@
-use dep::proof::{const::HASH_LENGTH, trie_proof::TrieProof, rlp::decode0, utils::memcpy};
-use dep::u2b::u64_to_u8;
+use dep::proof::{
+    const::{HASH_LENGTH, MAX_RECEIPT_RLP_LENGTH as MAX_ENCODED_RECEIPT_LENGTH}, trie_proof::TrieProof,
+    rlp::{RLP_List, decode1, decode0, STRING, LIST}, utils::{byte_value, memcpy}
+};
+use crate::receipt::{TxReceiptProof, TxReceiptPartial, MAX_RECEIPT_TREE_DEPTH, TxType, BLOOM_FILTER_LEN};
 
-use crate::receipt::{TxReceiptProof, TxReceiptPartial, MAX_RECEIPT_TREE_DEPTH};
-use crate::misc::{arrays::sub_array_equals_up_to_length, bytes::{nibbles_to_bytes, left_to_right_padding, right_pad}};
+global RECEIPT_FIELDS_COUNT = 4;
 
-global KEY_LEN = 6;
-global MAX_TX_IDX_BYTES_LEN = 2;
+global BYZANTIUM_BLOCK_NUM: u64 = 4_370_000;
+global MAX_RECEIPT_RLP_LENGTH: u64 = MAX_ENCODED_RECEIPT_LENGTH - 1;
 
-pub(crate) fn assert_receipt_equals<LOG_NUM, MAX_LOG_DATA_LEN, MAX_RECEIPT_RLP_LEN>(
-    block_number: Field,
-    receipt_rlp: [u8; MAX_RECEIPT_RLP_LEN],
-    receipt: TxReceiptPartial<LOG_NUM, MAX_LOG_DATA_LEN>
-) {
-    assert(false, "Not implemented");
+pub(crate) fn is_pre_byzantium(block_number: u64) -> bool {
+    block_number < BYZANTIUM_BLOCK_NUM
 }
 
-pub fn assert_tx_idx_equals(key_as_rlp: [u8; KEY_LEN], tx_idx: Field) {
-    let (key_right_padded, shifted_by) = left_to_right_padding(key_as_rlp);
-    let key_bytes: [u8; KEY_LEN / 2] = nibbles_to_bytes(key_right_padded);
-    let (rlp_offset, rlp_len) = decode0(key_bytes);
-    assert(shifted_by + 2 * (rlp_len + rlp_offset) == KEY_LEN, "key is not an rlp-encoded string");
-
-    if (rlp_len == 0) {
-        assert(tx_idx == 0, "key does not match rlp-encoded transaction index");
-    } else {
-        let mut tx_idx_bytes = [0; MAX_TX_IDX_BYTES_LEN];
-        let (tx_idx_right_padded, _) = right_pad(u64_to_u8(tx_idx as u64));
-        memcpy(&mut tx_idx_bytes, tx_idx_right_padded, 0);
-
+pub(crate) fn split_receipt_into_tx_type_and_receipt_rlp(
+    tx_type: TxType,
+    encoded_receipt: [u8; MAX_ENCODED_RECEIPT_LENGTH]
+) -> (u8, [u8; MAX_RECEIPT_RLP_LENGTH]) {
+    if (tx_type == 0) {
         assert(
-            sub_array_equals_up_to_length(tx_idx_bytes, key_bytes, rlp_offset, rlp_len), "key does not match rlp-encoded transaction index"
+            encoded_receipt[MAX_ENCODED_RECEIPT_LENGTH - 1] == 0, "Not 0x00 at the end of legacy receipt RLP"
```

Comment:
Or simpler even simpler: "Expected 0x00 at legacy receipt RLP"

Reply:
Assertion messages state what is wrong in given case. Like "Invalid...", "Wrong...". They don't say what should be. So I think that in this case it should also be written in that form.

Reply:
Changed to: "RLP receipt exceeds MAX_RECEIPT_RLP_LENGTH"
I feel like it's better because it actually tells you what's wrong on the higher level. Receipt is too long. Instead of focusing on details like having zero byte at the end